siriuspy/siriuspy/devices/injsys.py

```diff
@@ -771,3 +773,279 @@ def _command_target(self, cmdtype):
+        dpk_evt_ok = self.trigdpk.source == self.trignlk.source
+        dpk_dly_ok = self.trigdpk.delay_raw == self.SI_DPKCKR_DLYR_OPT
+        dpk_kck_ok = self.pudpk.strength == self.SI_DPKCKR_DEFKICK
+        dpk_on = (self.pudpk.pwrstate == PowerSupplyPU.PWRSTATE.On) and \
+            (self.pudpk.pulse == PowerSupplyPU.PULSTATE.On)
+        nlk_on = (self.punlk.pwrstate == PowerSupplyPU.PWRSTATE.On) and \
+            (self.punlk.pulse == PowerSupplyPU.PULSTATE.On)
+        return dpk_evt_ok & dpk_dly_ok & dpk_kck_ok & dpk_on & nlk_on
+
+    def cmd_switch_to_optim(self):
+        """Switch to Optimization."""
+        self._abort.clear()
+        self._update_status('Switching to PU Optimization config...')
+        if not self.pudpk.interlock_ok:
+            self._update_status('ERR: DpK interlocks not ok. Aborted.')
+            return False
+        if not self.punlk.interlock_ok:
+            self._update_status('ERR: NLK interlocks not ok. Aborted.')
+            return False
+        if self._abort.is_set():
+            self._clr_flag_abort()
+            return False
+
+        # if previously in on-axis, do delta angle x
+        if self.is_onaxis:
+            if self.posang.need_ref_update:
+                self.posang.cmd_update_reference()
+            _time.sleep(InjSysPUModeHandler._DEF_SLEEP)
+            delta = self.posang.delta_angx - self.TS_POSANG_DEFDELTA
+            self.posang.delta_angx = delta
+            if not self._wait(self.posang, 'delta_angx', delta):
+                self._update_status('ERR:Could not do delta AngX.')
+                return False
+
+        if self._abort.is_set():
+            self._clr_flag_abort()
+            return False
+
+        # set DpK trigger event to NLK trigger event
+        dpk_event = self.trignlk.source
+        self.trigdpk.source = dpk_event
+        if not self._wait(self.trigdpk, 'source', dpk_event):
+            self._update_status('ERR:Could not set DpK trigger event.')
+            return False
+        # set DpK trigger DelayRaw to 0
+        self.trigdpk.delay_raw = self.SI_DPKCKR_DLYR_OPT
+        if not self._wait(self.trigdpk, 'delay_raw', self.SI_DPKCKR_DLYR_OPT):
+            self._update_status('ERR:Could not set DpK trigger delay.')
+            return False
+
+        # set DpK Kick
+        self.pudpk.strength = self.SI_DPKCKR_DEFKICK
+        if not self._wait(self.pudpk, 'strength', self.SI_DPKCKR_DEFKICK):
+            self._update_status(
+                'ERR:Could not set DpK Kick to '
+                f'{self.SI_DPKCKR_DEFKICK:.1f}mrad.')
+            return False
+
+        # set pulsed magnet pwrstate and pulse
+        proced = (
+            (self.pudpk.cmd_turn_on, 'turn DpK on.'),
+            (self.punlk.cmd_turn_on, 'turn NLK on.'),
+            (self.pudpk.cmd_turn_on_pulse, 'turn DpK pulse on.'),
+            (self.punlk.cmd_turn_on_pulse, 'turn NLK pulse on.'),
+        )
+        for fun, msg in proced:
```

Comment:
maybe create a private method for this procedure, since it is repeated 3 times along the class

Reply:
Thanks @murilobalves , I pushed a commit trying to avoid repeated blocks of code. Done!